File: 2023/4/main.py
cards = [len(set.intersection(*[set(c.split()) for c in l.split(":")[1].split("|")])) for l in open(0).readlines()]
N = len(cards)

memo = [1] * N
for i in range(N):
    memo[i] += sum(memo[i - cards[N - i - 1] : i])
    
print(sum(memo))
for i, m in enumerate(memo):
    print(m, sum(memo[:i+1]))
    

File: 2023/4/main.jl
cards = [length(intersect([Set(split(c)) for c in split(split(l, ":")[2], "|")]...)) for l in readlines(stdin)]
N = length(cards)

memo = repeat([1], N)
for i = 1:N
    memo[i] += sum(memo[i - cards[(N + 1) - i] : i - 1])
end

println(sum(memo))
